controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+ * PWM outputs
+ *
+ * The following four outputs could be driven
+ * as PWM outputs:
+ *
+ * PA8  - Timer 1 Channel 1 - heater control
+ * PA11 - Timer 1 Channel 4 - solenoid
+ * PB3  - Timer 2 Channel 2 - blower control
+ * PB4  - Timer 3 Channel 1 - buzzer
+ *
+ * For now I'll just set up the blower since that's the only
+ * one called out in the HAL
+ *****************************************************************/
+static void InitPwmOut( void )
+{
+   // The PWM frequency isn't mentioned anywhere that I can find, so
+   // I'm just picking a reasonable number.  This can be refined later
+   const int pwmFreqHz  = 20000;
+
+   EnableClock( TIMER2_BASE );
+
+   // Connect PB3 to timer 2
+   GPIO_PinAltFunc( GPIO_B_BASE, 3, 1 );
+
+   TimerRegs *tmr = (TimerRegs *)TIMER2_BASE;
+
+   // Set the frequency
+   tmr->reload = (CPU_FREQ / pwmFreqHz)-1;
+
+   // Configure channel 2 in PWM output mode
+   tmr->ccMode[0] = 0x6800;
+
+   tmr->ccEnable = 0x10;
+
+   // Start with 0% duty cycle
+   tmr->compare[1] = 0;
+
+   // Load the shadow registers
+   tmr->event = 1;
+
+   // Start the counter
+   tmr->ctrl[0] = 0x81;
+}
+
+// Set the PWM period.  Currently the value is 0 to 255 which is
+// a shame.  We should change this to a float, say with a value 
+// of 0.0 to 1.0 to make better use of our resolution
+void HalApi::analogWrite(PwmPin pin, int value) {
+
+   // Convert the value to a float
+   float duty =  value * (1.0/255);
+
+   TimerRegs *tmr;
+   int chan;
+   switch( pin )
+   {
+      case PwmPin::BLOWER:
+         tmr = (TimerRegs *)TIMER2_BASE;
+         chan = 1;
+         break;
+
+      default:
+         return;
+   }
+
+   tmr->compare[chan] = tmr->reload * duty;
+}
+
+/******************************************************************
+ * Serial port to GUI
+ *****************************************************************/
+
+// This class is a generic circular buffer for byte sized data.
+// It could probably go in it's own header outside the HAL, we 
+// would just need to add interrupt suspend/restore to the HAL
+// definition.
+template <int N> class CircBuff
+{
+   uint8_t buff[N];
+   int head, tail;
+
+public:
+   CircBuff( void ){
+      head = tail = 0;
+   }
+
+   // Return number of bytes available in the buffer
+   int FullCt( void ){
+      int p = IntSuspend();
+      int ct = head-tail;
+      IntRestore(p);
+      if( ct < 0 ) ct += N;
+      return ct;
+   }
+
+   // Return number of free spaces in the buffer
+   int FreeCt( void ){
+      return N-1-FullCt();
+   }
+
+   // Get the oldest byte from the buffer.
+   // Returns -1 if the buffer is empty
+   int Get( void )
+   {
+      int ret = -1;
+
+      int p = IntSuspend();
+
+      if( head != tail )
+      {
+         ret = buff[tail++];
+         if( tail >= N ) tail = 0;
+      }
+
+      IntRestore( p );
+      return ret;
+   }
+
+   // Add a byte ot the buffer
+   // Retuns 1 on success, 0 if the buffer is full
```

Comment:
Return a bool then?

Reply:
It's the C coder in me showing through.  You guys and your fancy bool types!